server/apis/v1/handler.go

```diff
@@ -46,282 +68,831 @@ type handler struct {
 		return
 	}
-	c.JSON(http.StatusOK, isbsvc)
+	defer client.Close()
+
+	var (
+		minWM int64 = math.MaxInt64
+		maxWM int64 = math.MinInt64
+	)
+	watermarks, err := client.GetPipelineWatermarks(context.Background(), pipeline)
+	if err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to fetch pipeline: failed to calculate lag for pipeline %q: %s", pipeline, err.Error()))
+		return
+	}
+	for _, watermark := range watermarks {
+		// find the largest source vertex watermark
+		if _, ok := source[*watermark.From]; ok {
+			for _, wm := range watermark.Watermarks {
+				if wm > maxWM {
+					maxWM = wm
+				}
+			}
+		}
+		// find the smallest sink vertex watermark
+		if _, ok := sink[*watermark.To]; ok {
+			for _, wm := range watermark.Watermarks {
+				if wm < minWM {
+					minWM = wm
+				}
+			}
+		}
+	}
+	// if the data hasn't arrived the sink vertex
+	// use 0 instead of the initial watermark value -1
+	if minWM == -1 {
+		minWM = 0
+	}
+	lag = maxWM - minWM
+
+	pipelineResp := NewPipelineInfo(status, &lag, pl)
+	c.JSON(http.StatusOK, NewNumaflowAPIResponse(nil, pipelineResp))
 }
 
-// ListVertices is used to provide all the vertices of a pipeline
-func (h *handler) ListVertices(c *gin.Context) {
-	limit, _ := strconv.ParseInt(c.Query("limit"), 10, 64)
-	vertices, err := h.numaflowClient.Vertices(c.Param("namespace")).List(context.Background(), metav1.ListOptions{
-		LabelSelector: fmt.Sprintf("%s=%s", dfv1.KeyPipelineName, c.Param("pipeline")),
-		Limit:         limit,
-		Continue:      c.Query("continue"),
-	})
+// UpdatePipeline is used to update a given pipeline
+func (h *handler) UpdatePipeline(c *gin.Context) {
+	ns, pipeline := c.Param("namespace"), c.Param("pipeline")
+	// dryRun is used to check if the operation is just a validation or an actual update
+	dryRun := c.DefaultQuery("dry-run", "false")
+	dryRun = strings.ToLower(dryRun)
+
+	reqBody, err := parseSpecFromReq(c, SpecTypePipeline)
 	if err != nil {
-		c.JSON(http.StatusInternalServerError, err.Error())
+		h.respondWithError(c, fmt.Sprintf("Failed to parse request body, %s", err.Error()))
 		return
 	}
-	c.JSON(http.StatusOK, vertices.Items)
+
+	oldSpec, err := h.numaflowClient.Pipelines(ns).Get(context.Background(), pipeline, metav1.GetOptions{})
+	if err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to fetch pipeline %q namespace %q, %s", pipeline, ns, err.Error()))
+		return
+	}
+
+	updatedSpec, ok := reqBody.(*dfv1.Pipeline)
+	if !ok {
+		h.respondWithError(c, "Failed to convert request body to pipeline spec")
+		return
+	}
+	// Validate the namespace of the request
+	err = validateNamespace(h, updatedSpec, ns)
+	if err != nil {
+		h.respondWithError(c, err.Error())
+		return
+	}
+
+	// pipeline name in the URL should be same as spec name
+	if pipeline != updatedSpec.Name {
+		h.respondWithError(c, fmt.Sprintf("pipeline name %q is immutable", pipeline))
+		return
+	}
+
+	oldSpec.Spec = updatedSpec.Spec
```

Comment:
Logically, these 2 lines should be after the validation, right before the updating.